### PR TITLE
Pass HandshakeReq in all inbound lanes, #23527

### PR DESCRIPTION
**Pull request by patriknw**
The HandshakeReq message must be passed in each inbound lane to
ensure that it arrives before any application message. Otherwise there is a risk
that an application message arrives in the InboundHandshake stage before the
handshake is completed and then it would be dropped.

Refs #23527

### CODE REVIEW

File: akka-remote/src/main/scala/akka/remote/artery/BufferPool.scala [hunk 1]
```diff
@@ -37,7 +37,8 @@ private[remote] class EnvelopeBufferPool(maximumPayload: Int, maximumBuffers: In
     }
   }
 
-  def release(buffer: EnvelopeBuffer) = if (!availableBuffers.offer(buffer)) buffer.tryCleanDirectByteBuffer()
+  def release(buffer: EnvelopeBuffer) =
+    if (buffer.byteBuffer.isDirect && !availableBuffers.offer(buffer)) buffer.tryCleanDirectByteBuffer()
```
Comment (johanandren) [inline akka-remote/src/main/scala/akka/remote/artery/BufferPool.scala]: Is this an unrelated fix?

File: akka-remote/src/main/scala/akka/remote/artery/Codecs.scala [hunk 2]
```diff
@@ -649,3 +652,64 @@ private[remote] class Deserializer(
...
+            currentIterator = Vector.tabulate(numberOfLanes)(i ⇒ envelope.copyForLane(i)).iterator
+            push(out, currentIterator.next())
+          } finally {
+            val buf = envelope.envelopeBuffer
+            if (buf != null) {
+              envelope.releaseEnvelopeBuffer()
+              bufferPool.release(buf)
+            }
+          }
+        } else
```
Comment (johanandren) [inline akka-remote/src/main/scala/akka/remote/artery/Codecs.scala]: Could we replace the InHandler to one that doesn't need any condition at all after being done with the handshake instead?

File: akka-remote/src/main/scala/akka/remote/artery/Codecs.scala [hunk 3]
```diff
@@ -649,3 +652,64 @@ private[remote] class Deserializer(
...
+          push(out, envelope)
+      }
+
+      override def onPull(): Unit = {
+        if (currentIterator.isEmpty)
+          pull(in)
+        else {
+          push(out, currentIterator.next())
+          if (currentIterator.isEmpty) currentIterator = Iterator.empty // GC friendly
```
Comment (johanandren) [inline akka-remote/src/main/scala/akka/remote/artery/Codecs.scala]: Same here, the iterator never becomes non-empty again after becoming empty, right? So maybe both change handlers...

**Review by johanandren — APPROVED** [review]
LGTM, with one question and one possible (but maybe not necessary) optimization.

**Review by 2m — APPROVED** [review]
LGTM modulus Johan's comments.

File: akka-remote/src/main/scala/akka/remote/artery/BufferPool.scala (hunk 1 above)
Comment (patriknw) [reply akka-remote/src/main/scala/akka/remote/artery/BufferPool.scala]: no, when duplicating it is creating a new array based ByteBuffer and those should not be released back to the pool

File: akka-remote/src/main/scala/akka/remote/artery/Codecs.scala (hunk 2 above)
Comment (patriknw) [reply akka-remote/src/main/scala/akka/remote/artery/Codecs.scala]: It's not a final decision. The condition `envelope.association.isEmpty` may be true again after system restarts and then it's also good to do this. `envelope.association.isEmpty` should be cheap so this stage should not add much overhead.

**Comment by patriknw** [comment]
forgot mima

File: akka-remote/src/main/scala/akka/remote/artery/Codecs.scala (hunk 2 above)
Comment (johanandren) [reply akka-remote/src/main/scala/akka/remote/artery/Codecs.scala]: Ok, thanks for clarifying!

**Comment by patriknw** [comment]
Running in repeat job and found one possible related test failure:
```
[info] RemoteSendConsistencyWithThreeLanesSpec:
Coroner Thread Count starts at 44 in akka.remote.artery.RemoteSendConsistencyWithThreeLanesSpec
[info] Artery
# associate akka://systemB@localhost:33183
# associate akka://ArteryMultiNodeSpec@localhost:48124
# association setUID akka://ArteryMultiNodeSpec@localhost:48124#5466743159227436904
# association setUID akka://systemB@localhost:33183#3142185957702319102
[info] - must be able to identify a remote actor and ping it *** FAILED *** (5 seconds, 119 milliseconds)
[info]   java.lang.AssertionError: assertion failed: timeout (5 seconds) during expectMsgClass waiting for class akka.actor.ActorIdentity
[info]   at scala.Predef$.assert(Predef.scala:170)
[info]   at akka.testkit.TestKitBase$class.expectMsgClass_internal(TestKit.scala:508)
[info]   at akka.testkit.TestKitBase$class.expectMsgType(TestKit.scala:490)
[info]   at akka.testkit.TestKit.expectMsgType(TestKit.scala:882)
[info]   at akka.remote.artery.AbstractRemoteSendConsistencySpec$$anonfun$1$$anonfun$apply$mcV$sp$2.apply(RemoteSendConsistencySpec.scala:43)
[info]   at akka.remote.artery.AbstractRemoteSendConsistencySpec$$anonfun$1$$anonfun$apply$mcV$sp$2.apply(RemoteSendConsistencySpec.scala:33)
```

Will change to debug logging there.

**Review by ktoso — APPROVED** [review]
Fix looks good to me, LGTM
